server/src/main/java/com/cloud/user/AccountManagerImpl.java

```diff
@@ -1110,6 +1123,31 @@ public UserAccount createUserAccount(final String userName, final String passwor
         return _userAccountDao.findById(userId);
     }
 
+    /**
+     * if there is any permission under the requested role that is not permitted for the caller, refuse
+     */
+    private void checkRoleEscalation(Account caller, Long requestedRoleId) {
+        List<RolePermission> requestedPermissions = roleService.findAllPermissionsBy(requestedRoleId);
+
+        for (RolePermission permission : requestedPermissions) {
+            if (permission.getPermission() == Permission.ALLOW) {
+                String command = permission.getRule().getRuleString();
+                if (command.contains("*")) {
+                    // find all api that match command and loop
+                    checkApiAccess(caller, command);
+                } else {
+                    checkApiAccess(caller, command);
+                }
```

Comment:
maybe this if-else clause isn't required?

Reply:
:) it will be when done, still in progress @Pearl1594